### ISSUE TITLE
Fixup: Use expanding ranges

### ISSUE THREAD

**Issue by umpox** [sourcegraph/cody#585]
## Description

Currently fixups will use the users selected range, if it is available.

This is _ok_ but it becomes problematic in different scenarios.

Example:

Code actions -> Do we really just need to change the code that has the error? Often we need to change much more to satisfy the error.

Refactors -> I want to change the returned value of a function, if I have only selected the return value then we should expand the range scope to include the entire function.

## Implementation

I think we should investigate what the best option here is. We should look at competitors like Copilot too.

One basic idea would just be to always expand the range scope to the nearest function (if available).

**Comment by arafatkatze** [comment]
At a cursory out look I saw that cody does use Prefix/Postfix text  in the comprehension of Fixup recipes. 

https://github.com/sourcegraph/cody/blob/54db16dc212702ee20fb50436c02ad8849eccab7/vscode/src/non-stop/FixupController.ts#L283-L295

And even for FixupCodeAction 

https://github.com/sourcegraph/cody/blob/54db16dc212702ee20fb50436c02ad8849eccab7/vscode/src/code-actions/fixup.ts#L22-L28

The broader question is that while the range expansion for input is good. How to extend it intelligently to the part where the fixup is actually applied. Like should it only be applied to EXACTLY what the user selected and if we wanna do more. Is it possible we might cause inconvenience for the user by going out of bounds?

**Comment by dominiccooney** [comment]
This is a great idea, we should do it.

FWIW, anecdotally from when I built this:
- Fixups performs better with the before/after context. (Unsurprising... so would a human.)
- Fixups performs poorly when it would be beneficial to expand the scope. For example, you ask it to rename something but the declaration is outside the scope being edited. Typical bad behavior: repeating code outside the selection (which IIRC we now prompt to suppress); not generating output; etc.

In the first iteration I tried to let Cody push back in chat if you did that, but it was way too chatty in general so @abeatrix silenced it.

Expanding the selection to a boundary which trades off logical sense (e.g. "whole function") and amount of code sent sounds great as a first cut.

Performance can deteriorate asking it to reiterate a lot of code... we could relax the requirement to repeat all the code and use a diff to align the output we *do* get to the content of the file.

Hard mode: Edit multiple files/places. Might be interesting to let Cody pitch toil to the editor, for example, have it explain a multi-step refactoring with relatively abstract commands like "rename variable x to y" and then we do those precisely.
